platforms/core-configuration/model-core/src/integTest/groovy/org/gradle/api/provider/MergeProviderIntegrationTest.groovy

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.provider
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+/**
+ * Tests {@link org.gradle.api.internal.provider.MergeProvider}.
+ */
+class MergeProviderIntegrationTest extends AbstractIntegrationSpec {
+
+    def "carries task dependencies"() {
```

Comment:
Doesn't `flatMap` carry the task dependencies of the original provider?

Reply:
Also if it didn't, wouldn't that just be testing the behavior of `flatMap` and not this provider?

Reply:
Right, it's just confirming that if you offer `flapMap`s (or mix `map`s and `flapMap`s) as your source Providers in the arguments to the `MergeProvider`, it properly doesn't carry the task deps from `flatMap`.

Reply:
I guess my point is, all that does is test the behavior of `flatMap` in terms of how _it_ tracks task dependencies. `MergeProvider` treats all providers equivalently, so passing some `flatMap` provider would be equivalent to passing in a `Providers.notDefined` and testing its task dependencies

Reply:
Right, that's exactly the behavior I want to make explicit in a test.  

From an outsider's view, I have no way of knowing to expect `MergeProvider` to honor or discard the task dependencies from any of its source `Providers`.  There's no notes on the class - I'd have to read through the source code or experiment with it.

This existing test confirms it honors `map` that carries task deps.  No test confirms the behavior of `flatMap` or a mix of both types.  It's a reasonable (but wrong) to maybe expect a mix wouldn't keep **any** of the task deps around.

Reply:
But the behavior of whether `map` or `flatMap` passes through task dependencies is implemented by `TransformBackedProvider` and `FlatMapProvider`. We should not be testing whether those providers work in this test

I would rather remove `map` from this test rather than add `flatMap`, as regardless of the type of provider passed to the constructor, the behavior should not change